src/lightning/fabric/plugins/precision/fp8_transformer_engine.py

```diff
@@ -0,0 +1,181 @@
+    <https://docs.nvidia.com/deeplearning/transformer-engine`__.
+
+    .. warning::  This is an :ref:`experimental <versioning:Experimental API>` feature.
+
+    Args:
+        dtype: The base dtype to use.
+        recipe: Recipe for the DelayedScaling
+            `configuration <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transform
+            er_engine.common.recipe.DelayedScaling`__. In dict format or the dataclass format.
+        replace_layers: Whether to replace ``Linear`` and ``LayerNorm`` layers automatically with their Transformer
+            Engine alternatives. Note that they don't subclass the torch equivalents so checks like
+            ``isinstance(l, torch.nn.Linear)`` will not pass.
+
+    .. note::
+
+        Support for FP8 in the linear layers with `precision='8-mixed'` is currently limited to tensors with shapes
+        where the dimensions are divisible by 8 and 16 respectively. You might want to add padding to your inputs to
+        conform to this restriction.
+    """
+
+    precision: Literal["8-mixed-transformer-engine"] = "8-mixed-transformer-engine"
+
+    def __init__(
+        self,
+        dtype: Optional[torch.dtype] = None,
+        recipe: Optional[Union[Mapping[str, Any], "DelayedScaling"]] = None,
+        replace_layers: bool = True,
+    ) -> None:
+        if not _TRANSFORMER_ENGINE_AVAILABLE:
+            raise ModuleNotFoundError(str(_TRANSFORMER_ENGINE_AVAILABLE))
+        from transformer_engine.common.recipe import DelayedScaling
+
+        if recipe is None:
+            recipe = DelayedScaling()
+        elif isinstance(recipe, Mapping):
+            recipe = dict(recipe)  # copy
+            if "fp8_format" in recipe:
+                from transformer_engine.common.recipe import Format
+
+                recipe["fp8_format"] = getattr(Format, recipe["fp8_format"])
+            recipe = DelayedScaling(**recipe)
+
+        if dtype is None:
+            dtype = torch.get_default_dtype()
+        self.dtype = dtype
+        self.recipe = recipe
+        self.replace_layers = replace_layers
+
+    @contextmanager
+    def forward_context(self) -> Generator[None, None, None]:
+        default_dtype = torch.get_default_dtype()
+        torch.set_default_dtype(self.dtype)
+
+        import transformer_engine.pytorch as te
+
+        with te.fp8_autocast(enabled=True, fp8_recipe=self.recipe):
+            yield
+
+        torch.set_default_dtype(default_dtype)
+
+    def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
+        # avoid converting if any is found. assume the user took care of it
+        if self.replace_layers and not any("transformer_engine" in m.__module__ for m in module.modules()):
+            _convert_layers(module, self.dtype)
+        module = module.to(dtype=self.dtype)
+        return module
+
+    @contextmanager
+    def init_context(self) -> Generator[None, None, None]:
```

Comment:
what does the tripple none as yiled means?

Reply:
See https://docs.python.org/3/library/typing.html#typing.Generator